### PR TITLE
Normalize merge-tree's segment order on rebase of an insert op

### PR DESCRIPTION
**Pull request by Abe27342**
## Description

This fixes an eventual consistency issue exposed by fuzz testing.
The core problem is that clients that are reconnecting may end up in a state in which they don't resolve tiebreaks of inserted segments the same way as other clients.
This is because the tiebreaking logic makes assumptions about the relative ordering of inserted/removed segments in the collab window.
Specifically, in the absence of reconnect it's never the case that a segment inserted at seq A appears immediately after a segment removed at seq B, where:
- both A and B in the collab window
- A > B
- the ops with seq A and seq B are concurrent

(In this case, the insertion logic would have placed the inserted segment before the removed segment on all clients).

However, this invariant was no longer true with reconnect due to rebasing ops.
The regression test demonstrates a case where this happens.
The proposed fix implemented in this PR is to have the merge-tree update its internal state when one of its pending segment groups is rebased to instead be in terms of the current sequence number.

[AB#1675](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/1675)

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -2138,6 +2125,118 @@ export class MergeTree {
...
+        const currentOrder = [...adjacentRemovedSegs, ...consecutiveLocalSegments].map(
+            (seg) => ({
+                parent: seg.parent,
+                index: seg.index,
+                ordinal: seg.ordinal,
+            }),
+        );
+
+        const newOrder = [...consecutiveLocalSegments, ...adjacentRemovedSegs];
+        // TODO: Invoke slide ref callback and test that this happens
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/mergeTree.ts:2211]: I wanted to merge in Tony's walk references helper for this

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2211]: it's in now

File: packages/dds/merge-tree/src/test/client.applyMsg.spec.ts [hunk 2]
```diff
@@ -539,56 +546,35 @@ _: Local State
...
+7:6:B0@6   | DDDDDDb ---D  | 7:6:B0@6   | DDDDDDb- -_-D
+           |               |            |           c
+8:6:C0@6   | DDDDDDcb ---D | 8:6:C0@6   | DDDDDDb- -c-D
 Client C does not match client A
 ```
      */
-    it.skip("Concurrent insert into removed segment across block boundary", () => {
+    it("Concurrent insert into removed segment across block boundary", () => {
         const clients = createClientsAtInitialState(
             { initialState: "", options: { mergeTreeUseNewLengthCalculations: true } },
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/client.applyMsg.spec.ts:574]: how does this behave without mergeTreeUseNewLengthCalculations?

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 3]
```diff
@@ -2138,6 +2125,118 @@ export class MergeTree {
...
+     *
+     * The difference in ordering can be problematic for tie-breaking concurrently inserted segments in some scenarios.
+     * Rather than incur extra work tie-breaking these scenarios for all clients, when client 1 rebases its operation,
+     * it can fix up its local state to align with what would be expected of the op it resubmits.
+     */
+    public normalizeSegmentsOnRebase(segmentGroup: SegmentGroup): void {
+        // Forward and backward excursions will expand the affected normalization range,
+        // so simply taking the first segment suffices
+        const { refSeq, segments: [segment] } = segmentGroup;
+        const forwardLocalSegs: ISegment[] = [];
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:2169]: could we get by with a single list here, insert the original seg, and unshift backward, and push forward, then we don't have to enumerate and build another array to get to consecutiveLocalSegments

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 4]
```diff
@@ -2138,6 +2125,118 @@ export class MergeTree {
...
+        const adjacentRemovedSegs: ISegment[] = [];
+        backwardExcursion(segment, (seg) => {
+            if (seg.seq === UnassignedSequenceNumber) {
+                backwardLocalSegs.push(seg);
+                return true;
+            }
+            const removalInfo = toRemovalInfo(seg);
+            if (removalInfo !== undefined &&
+                removalInfo.removedSeq !== UnassignedSequenceNumber &&
+                removalInfo.removedSeq > refSeq
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:2180]: minseq?

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 5]
```diff
@@ -2138,6 +2125,118 @@ export class MergeTree {
...
+
+        const currentOrder = [...adjacentRemovedSegs, ...consecutiveLocalSegments].map(
+            (seg) => ({
+                parent: seg.parent,
+                index: seg.index,
+                ordinal: seg.ordinal,
+            }),
+        );
+
+        const newOrder = [...consecutiveLocalSegments, ...adjacentRemovedSegs];
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2210]: also, this doesn't seem idempotent, like what happens if there are multiple local inserts? 

would it make more sense to just rebase everything at once in some kind of fix up pass before regenerate.

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2180]: actually, i think you only want seen removes. the computation is more complicated, as you also need to look a client ids. easiest to just get the length, and differentiate based on 0 or undefined.

if we do some global pass, we could also think about just running zamboni to clean up as much as possible, then all undefined length segments will be gone, and should keep you decoupled from mergeTreeUseNewLengthCalculations.

not sure how reasonable it would be, but possibly the rebase could be part of Zamboni so we just get the string into good shape overall, including ordering fix ups. definitely not necessary, but maybe consider it longer term

**Comment by anthony-murphy** [comment]
a thought i had on this area, could we augment the reconnect farm to hit it? that test is pretty simplistic right now.

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Abe27342) [reply packages/dds/merge-tree/src/mergeTree.ts]: Pushed an approach which I think is nicer that's more along these lines. I'm no longer seeing any fuzz failures on the reconnect tests for interval collection if I modify the operations to not include intervals (so it's just basic add/remove text), whereas previously there was a ~2.5% failure rate over 1k tests.

I was mistaken in Standup the other day, fixing up the callback invocation on local references doesn't fix interval collections stuff. But the core of that problem is that the reconnect strategy used for interval collection requires rebasing character positions from a previous localSeq, which normalizing the segment order upon reconnect completely messes up. I think a reasonable enough way to solve that might be for SharedSegmentSequence to give interval collection a chance to query the underlying merge-tree pre-normalization, and then IntervalCollection's pending state tracking can store the character position + refSeq/localSeq combos it cares about.

I also like the idea to modify merge-tree reconnect farm to repro this issue, I'll do that in this PR before checkin.

**Comment by Abe27342** [comment]
> a thought i had on this area, could we augment the reconnect farm to hit it? that test is pretty simplistic right now.

Changing it to have both a client that always disconnects on application and a different one that only sometimes disconnects seems to suffice to make it hit the issue. There are still some failures on the long version of that test after doing so, but ~80% less (and not on new tests), so I'm reasonably comfortable checking this in.

File: packages/dds/merge-tree/src/test/client.applyMsg.spec.ts (hunk 2 above)
Comment (Abe27342) [reply packages/dds/merge-tree/src/test/client.applyMsg.spec.ts:577]: it passes with and without.

File: packages/dds/merge-tree/src/client.ts [hunk 6]
```diff
@@ -67,7 +67,15 @@ function elapsedMicroseconds(trace: Trace) {
...
 }
 
-export class Client {
+/**
+ * @internal
+ * Emitted before this client's merge-tree normalizes its segments on reconnect, potentially
+ * ordering them. Useful for DDS-like consumers built atop the merge-tree to compute any information
+ * they need for rebasing their ops on reconnection.
+ */
+export type IClientEvents = (event: "normalize", listener: () => void) => void;
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/client.ts:76]: i'm not against Client being an event emitter, but it doesn't align with the existing callback pattern, and i prefer we are consistent. Also probably something to do in another pr.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 7]
```diff
@@ -2133,6 +2122,173 @@ export class MergeTree {
...
+                true,
+            );
+            this.localPartialsComputed = true;
+        }
+    }
+
+    // Segments should either be removed remotely, removed locally, or inserted locally
+    private normalizeAdjacentSegments(affectedSegments: List<ISegment>): void {
+        const currentOrderTest = Array.from(affectedSegments.map(({ data }) => data));
+        // Eagerly demand this since we're about to shift elements in the list around
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:2145]: seems like these two lines could be combined and reduce an array allocation and enumeration

File: packages/dds/merge-tree/src/client.ts (hunk 6 above)
Comment (Abe27342) [reply packages/dds/merge-tree/src/client.ts:76]: yeah... I hated that inconsistency while I was writing it. The slightly awkward thing is that you can only attach one callback, but we could always work around that by making interval collections' subscription check for an existing callback and call it, effectively making a linked list. It's a bit icky but I don't think it'll be an issue as there probably aren't a ton of interval collections on any given shared string.

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/client.ts:76]: ahh. i didn't notice the multiple listeners, i only saw one in interval collection. i'd also be fine with openning an issue to move the other callbacks to events

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/client.ts:76]: just re-checked and there is only one listener? maybe one got removed in refactoring? still open to changing later, as i do something similar wrangling is tests to get multiple listeners

Comment (Abe27342) [reply packages/dds/merge-tree/src/client.ts:76]: the thing is each interval collection on the SharedString needs to attach to the underlying merge-tree event. There can be multiple. Alternatively we could have SharedSegmentSequence subscribe and invoke an API on IntervalCollection, but that seems more gross imo (users actually get their hands on IntervalCollection whereas they generally don't get their hands on a Client)

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 7 above)
Comment (Abe27342) [reply packages/dds/merge-tree/src/mergeTree.ts:2144]: oops yeah I meant to fix this, good catch. Using the second param on Array.from() should reduce to only one array allocation..

File: packages/dds/merge-tree/src/client.ts (hunk 6 above)
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/client.ts:77]: i'm fine with this. lets just open an issue to move client to the event emitter model. there are some test cases this will simplify as well

File: packages/dds/merge-tree/src/client.ts [hunk 8]
```diff
@@ -67,7 +67,16 @@ function elapsedMicroseconds(trace: Trace) {
...
 
-export class Client {
+/**
+ * Emitted before this client's merge-tree normalizes its segments on reconnect, potentially
+ * ordering them. Useful for DDS-like consumers built atop the merge-tree to compute any information
+ * they need for rebasing their ops on reconnection.
+ *
+ * @internal
+ */
+export type IClientEvents = (event: "normalize", listener: () => void) => void;
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/client.ts:77]:
```suggestion
export type IClientEvents = (event: "normalize", listener: (target: IEventThisPlaceHolder) => void) => void;
```

i like to give the object to the event handler. this makes event handler re-usable without the need to create new function objects for each client that uses it

File: packages/dds/merge-tree/src/test/mergeTreeOperationRunner.ts [hunk 9]
```diff
@@ -161,22 +161,22 @@ export function generateOperationMessagesForClients(
...
         // pick a client greater than 0, client 0 only applies remote ops
         // and is our baseline
-        const client = clients[random.integer(1, clients.length - 1)(mt)];
+        const client = clients[random.integer(1, clients.length - 1)];
         const len = client.getLength();
         const sg = client.peekPendingSegmentGroups();
         let op: IMergeTreeOp | undefined;
-        if (len === 0 || len < minLength) {
-            const text = client.longClientId!.repeat(random.integer(1, 3)(mt));
+        if (len === 0 || len < minLength || (len < minLength * 3 && random.bool(1 / 3))) {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/mergeTreeOperationRunner.ts:168]: why add the insert condition? can't the test just add an insert operation if it wants it?

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/test/mergeTreeOperationRunner.ts:168]: this is just meant to be a fallback case, that handles empty/min len, as we usually want data in the tree.

**Review by anthony-murphy — APPROVED** [review]
a couple small comments, but otherwise looks good.

File: packages/dds/merge-tree/src/test/mergeTreeOperationRunner.ts (hunk 9 above)
Comment (Abe27342) [reply packages/dds/merge-tree/src/test/mergeTreeOperationRunner.ts:168]: yeah, that seems better. I'll convert tests that seem like they should have an insert op to inject one in config. The lack of inserts was one reason the reconnect farm didn't repro this issue (text was only inserted when the string dipped below min length, so it was pretty unlikely to have a lot of adjacent surviving segments that would end up in a different order on rebase)

File: packages/dds/merge-tree/src/client.ts (hunk 6 above)
Comment (Abe27342) [reply packages/dds/merge-tree/src/client.ts]: Sounds good. Filed AB#2224 for this.
